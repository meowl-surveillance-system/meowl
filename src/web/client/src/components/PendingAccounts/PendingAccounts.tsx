import React, { Component } from "react";
import { Container, Typography } from "@material-ui/core";
import List from "@material-ui/core/List";
import ListItem from "@material-ui/core/ListItem";
import ListSubheader from "@material-ui/core/ListSubheader";
import Button from "@material-ui/core/Button";

interface Props {}
interface State {
  selectedAccount: string;
  accountList: Array<string>;
}

/**
 * A component for viewing and approving pending accounts
 */
export default class PendingAccounts extends Component<Props, State> {
  constructor(props: Props) {
    super(props);
    this.state = {
      selectedAccount: "",
      accountList: [],
    };
  }

  /**
   * Fetch the pendingAccounts from backend
   */
  componentDidMount() {
    fetch(`/auth/getPendingAccounts`)
      .then((res) => res.json())
      .then((usernames) => this.setState({ accountList: usernames }))
      .catch((e) => console.error(e));
  }

  /**
   * Renders the list of pending accounts
   */
  renderList = () => {
    return this.state.accountList.map((username, index) => {
      return (
        <ListItem
          button
          key={index}
          onClick={() => this.setState({ selectedAccount: username })}
        >
          {username}
        </ListItem>
      );
    });
  };

  /**
   * Rejects the selectedAccount
   */
  rejectAccount = (): void => {
    const requestOptions = {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        username: this.state.selectedAccount,
      }),
    };
    fetch(`/auth/rejectRegistration`, requestOptions)
      .then((res) => res.text())
      .then((msg) => {
        alert(msg);
      })
      .catch((e) => console.log(e));
  };

  /**
   * Approves the selectedAccount
   */
  approveAccount = (): void => {
    const requestOptions = {
      method: "POST",
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      body: JSON.stringify({
        username: this.state.selectedAccount,
      }),
    };
    fetch(`/auth/approveRegistration`, requestOptions)
      .then((res) => res.text())
      .then((msg) => {
        alert(msg);
      })
      .catch((e) => console.error(e));
  };

  /**
   * Renders out the approve and reject buttons and a list of accounts
   */
  render() {
    return (
      <Container>
        <Button
          variant="contained"
          color="secondary"
          onClick={() => this.rejectAccount()}
        >
          Reject
        </Button>
        <Button
          variant="contained"
          color="primary"
          onClick={() => this.approveAccount()}
        >
          Approve
        </Button>
        <List
          subheader={
            <ListSubheader color="inherit">
              <Typography variant="inherit">Pending Accounts</Typography>
            </ListSubheader>
          }
        >
          {this.renderList()}
        </List>
      </Container>
    );
  }
}
