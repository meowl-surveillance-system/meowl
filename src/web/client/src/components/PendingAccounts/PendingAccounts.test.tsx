import React from "react";
import { shallow } from "enzyme";
import { render } from "@testing-library/react";
import PendingAccounts from "./PendingAccounts";

describe("PendingAccounts component", () => {
  let mockArray: Array<string>;
  let mockJsonPromise: Promise<Array<string>>;
  let mockFetchPromise: Promise<object>;
  beforeEach(() => {
    mockArray = ["hooomygod", "endme"];
    mockJsonPromise = Promise.resolve(mockArray);
    mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise,
    });
    jest
      .spyOn(window, "fetch")
      .mockImplementationOnce(() => mockFetchPromise as Promise<Response>);
  });

  it("renders the PendingAccounts component", () => {
    const { container } = render(<PendingAccounts />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("should reject selected account", () => {
    const mockState: object = {
      selectedAccount: mockArray[0],
    };
    const wrapper: any = shallow(<PendingAccounts />);
    wrapper.instance().setState(mockState);
    const mockTextPromise = Promise.resolve("testMsg");
    const mockRejectPromise = Promise.resolve({
      text: () => mockTextPromise,
    });
    jest
      .spyOn(window, "fetch")
      .mockImplementationOnce(() => mockRejectPromise as Promise<Response>);
    jest.spyOn(window, "alert").mockImplementationOnce(() => {});
    wrapper.instance().rejectAccount();
    const mockRequest = {
      body: '{"username":"' + wrapper.instance().state.selectedAccount + '"}',
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      method: "POST",
    };
    expect(window.fetch).toHaveBeenCalledWith(
      `/auth/rejectRegistration`,
      mockRequest,
    );
  });

  it("should approve selected account", () => {
    const mockState: object = {
      selectedAccount: mockArray[1],
    };
    const wrapper: any = shallow(<PendingAccounts />);
    wrapper.instance().setState(mockState);
    const mockTextPromise = Promise.resolve("testMsg");
    const mockApprovePromise = Promise.resolve({
      text: () => mockTextPromise,
    });
    jest
      .spyOn(window, "fetch")
      .mockImplementationOnce(() => mockApprovePromise as Promise<Response>);
    jest.spyOn(window, "alert").mockImplementationOnce(() => {});
    wrapper.instance().approveAccount();
    const mockRequest = {
      body: '{"username":"' + wrapper.instance().state.selectedAccount + '"}',
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
      },
      method: "POST",
    };
    expect(window.fetch).toHaveBeenCalledWith(
      `/auth/approveRegistration`,
      mockRequest,
    );
  });
});
