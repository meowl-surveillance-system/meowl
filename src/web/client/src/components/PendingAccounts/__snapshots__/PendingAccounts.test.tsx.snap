// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PendingAccounts component renders the PendingAccounts component 1`] = `
<div
  class="MuiContainer-root MuiContainer-maxWidthLg"
>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Reject
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <button
    class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary"
    tabindex="0"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      Approve
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
  <ul
    class="MuiList-root MuiList-padding MuiList-subheader"
  >
    <li
      class="MuiListSubheader-root MuiListSubheader-colorInherit MuiListSubheader-sticky MuiListSubheader-gutters"
    >
      <span
        class="MuiTypography-root"
      >
        Pending Accounts
      </span>
    </li>
  </ul>
</div>
`;
